### PR TITLE
Define methods for big(T)

### PR DESCRIPTION
**Pull request by giordano**
This PR defines methods for `big(T)`, which return an appropriate type to represent a value of type `T` as an arbitrary precision number.  I believe this is useful *per se* and also to define a single simple method to promote a complex number to arbitrary precision, as suggested [here](https://github.com/JuliaLang/julia/issues/21204#issuecomment-289910964).  If this PR will be merged, I'll follow up with a patch implementing `big{T<:Real}(z::Complex{T})`.

### CODE REVIEW

**Review by andreasnoack — APPROVED** [review]

**Comment by StefanKarpinski** [comment]
@JeffBezanson has sometimes had objections to functions that work in both the type and value domain, but this one seems harmless enough to me – hard to imagine what the ambiguity would be.

**Review by ararslan — APPROVED** [review]

**Comment by StefanKarpinski** [comment]
Note that this needs to wait until after 0.6 is branched to be merged.

**Comment by giordano** [comment]
Rebased on master, used new `where` syntax, added the method
```julia
big(z::Complex{T}) where {T<:Real} = Complex{big(T)}(z)
```
that replaces the two methods
```julia
big(z::Complex{<:AbstractFloat}) = Complex{BigFloat}(z)
big(z::Complex{<:Integer}) = Complex{BigInt}(z)
```
as suggested by @ararslan at https://github.com/JuliaLang/julia/issues/21204#issuecomment-289910964.

File: base/complex.jl [hunk 1]
```diff
@@ -880,8 +880,8 @@ end
 float(z::Complex{<:AbstractFloat}) = z
 float(z::Complex) = Complex(float(real(z)), float(imag(z)))
 
-big(z::Complex{<:AbstractFloat}) = Complex{BigFloat}(z)
-big(z::Complex{<:Integer}) = Complex{BigInt}(z)
+big(::Type{Complex{T}}) where {T} = Complex{big(T)}
```
Comment (ararslan) [inline base/complex.jl]: It might make sense to impose `T<:Real` here.

Comment (giordano) [inline base/complex.jl]: Done, thanks.

File: base/number.jl [hunk 2]
```diff
@@ -211,3 +211,22 @@ julia> factorial(big(21))
 ```
 """
 factorial(x::Number) = gamma(x + 1) # fallback for x not Integer
+
+"""
+    big(T::Type)
+
+Returns an appropriate type to represent a value of type `T` as an arbitrary precision number.
```
Comment (ararslan) [inline base/number.jl]: I'd word this like
> Compute the type that represents the numeric type `T` with arbitrary precision.

Comment (giordano) [inline base/number.jl]: Reworded, thanks again for the suggestion!

**Review by ararslan — APPROVED** [review]
Nice work!

**Review by StefanKarpinski — APPROVED** [review]

**Pull request merged**